roachprod: garbage collect DNS records

herkolategan opened the pull request:
The commits in this PR pertain to deleting DNS records through garbage collection. The functionality has been removed from `DestroyCluster`. Moving the functionality the the garbage collection job ensures that all dangling DNS records will eventually be deleted.

Release Note: None
Epic: None

herkolategan committed:
roachprod: list and delete DNS records

Previously the Provider DNS API did not expose a way for all records in a zone
to be listed, or to delete a list of records. This change adds the ability to
list all DNS records and to delete a set of records by name. This provides a way
for `roachprod` GC to be able to find dangling DNS records and delete it.

Additionally, a few DNS functions have also been updated to take a context and
pass it along to the execution commands.

herkolategan committed:
roachprod: garbage collect DNS records

Previously we relied on destroy cluster to delete DNS records. This has been
removed in a previous commit. This responsibility has been moved to garbage
collection. All DNS records from each DNS Provider is listed and if any DNS
records no longer correspond to a live cluster it will be deleted.

herkolategan committed:
roachprod: add docs for DNSProvider interface

Document the functions on the DNSProvider interface to make it clear what the
implementations of the interface should adhere to.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/111260)
<!-- Reviewable:end -->

srosenberg commented on pkg/roachprod/roachprod.go:
```diff
@@ -1413,18 +1413,26 @@ func Create(
+		clusterErr := cloud.GCClusters(l, cld, dryrun)
+		combinedErrors = errors.CombineErrors(combinedErrors, clusterErr)
+		// GCDNS depends on ListCloud so only call it if ListCloud runs without errors.
+		dnsErr := cloud.GCDNS(l, cld, dryrun)
```
Should this be run in a separate goroutine?

renatolabs commented on pkg/roachprod/cloud/cluster_cloud.go:
```diff
@@ -280,19 +280,13 @@ func CreateCluster(
-		return err
-	}
-	return vm.FanOutDNS(c.VMs, func(p vm.DNSProvider, vms vm.List) error {
-		return p.DeleteRecordsBySubdomain(c.Name)
```
Do we _need_ to remove this? We can leave GC to remove dangling resources (such as when the roachtest process crashes).

Also, what happens to local clusters?

herkolategan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111260)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @renatolabs, and @srosenberg)

___
*[`pkg/roachprod/roachprod.go` line 1425 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/111260#gh-1340200176:-NfSfId_-GZ5mOEnpbCe:b-kohu0p) ([raw file](https://github.com/cockroachdb/cockroach/blob/22cab512b4fd0538cd1786d24f05421912555ac3/pkg/roachprod/roachprod.go#L1425)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

It seems like if this `err` is not nil, it's never logged or included in `combinedErrors`.
</blockquote></details>

Seems is an understatement, this one slipped by me! Thanks for catching!
For some reason I always think we have a linter to capture simple cases like these.
___
*[`pkg/roachprod/roachprod.go` line 1430 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/111260#gh-1339392059:-NfSf_zK5idp6rhzKjlN:b7w5nl5) ([raw file](https://github.com/cockroachdb/cockroach/blob/22cab512b4fd0538cd1786d24f05421912555ac3/pkg/roachprod/roachprod.go#L1430)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

Should this be run in a separate goroutine?
</blockquote></details>

It can and the idea did cross my mind. It's probably worth while.


<!-- Sent from Reviewable.io -->

herkolategan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111260)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @renatolabs, and @srosenberg)

___
*[`pkg/roachprod/cloud/cluster_cloud.go` line 0 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/111260#gh-1340588680:-NfShsua-hdZ_SDyLXg_:b-nvjjge) ([raw file](https://github.com/cockroachdb/cockroach/blob/22cab512b4fd0538cd1786d24f05421912555ac3/pkg/roachprod/cloud/cluster_cloud.go#L0)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

*(Reviewable was [unable](https://github.com/Reviewable/Reviewable/issues/14) to map this GitHub inline comment thread to the right spot — sorry!)*

Do we _need_ to remove this? We can leave GC to remove dangling resources (such as when the roachtest process crashes).

Also, what happens to local clusters?
</blockquote></details>

`local cluster` does have its own clean-up section that removes the dns records from the local dns cache. I was in two minds for having the deletion in two places, since the cost is low and the records do little harm if they remain for a while. But I can add back the code, it probably makes more sense to try and do the deletion immediately.


<!-- Sent from Reviewable.io -->

herkolategan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111260)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @renatolabs, and @srosenberg)

___
*[`pkg/roachprod/cloud/cluster_cloud.go` line 0 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/111260#gh-1340588680:-NfSiW4-0SuXndyqhgyk:b-y679dv) ([raw file](https://github.com/cockroachdb/cockroach/blob/22cab512b4fd0538cd1786d24f05421912555ac3/pkg/roachprod/cloud/cluster_cloud.go#L0)):*
<details><summary><i>Previously, herkolategan (Herko Lategan) wrote…</i></summary><blockquote>

`local cluster` does have its own clean-up section that removes the dns records from the local dns cache. I was in two minds for having the deletion in two places, since the cost is low and the records do little harm if they remain for a while. But I can add back the code, it probably makes more sense to try and do the deletion immediately.
</blockquote></details>

I do want to however move the deletion before the VMs are destroy to avoid a race where GC could find there are no VMs / cluster and tries to clean up dangling DNS records while the section might also try to execute here.


<!-- Sent from Reviewable.io -->

herkolategan committed:
roachprod: destroy DNS entries first

Previously DNS records were deleted after a cluster is destroyed. In order to
avoid an improbable race condition with the garbage collection job DNS records
will now be deleted before the VMs are deleted.

If the VMs were deleted first it would be possible that the GC job no longer
detects the cluster and also tries to delete the DNS records while a cluster is
also being destroyed.

Release Note: None
Epic: None

renatolabs reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111260#-:-NfW2TnV-da-0P1Sdfah:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @DarrylWong and @srosenberg)



<!-- Sent from Reviewable.io -->

srosenberg reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

It'd be nice to add some tests, but that's really blocked but not having an infra. for integration tests; unit tests won't really do much here other than smoke testing. Since it's not a critical path, manual test runs should suffice for now.

> Reviewed 3 of 4 files at r2, 5 of 5 files at r5, 4 of 4 files at r6, 2 of 2 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111260#-:-Nfrs0c7CKhXtLXeJu8F:b-s2pskv)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @DarrylWong, @herkolategan, and @renatolabs)

___
*[`pkg/roachprod/roachprod.go` line 1430 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/111260#gh-1339392059:-NfrqdEsC5TRNi4-4lbC:b-xyly66) ([raw file](https://github.com/cockroachdb/cockroach/blob/22cab512b4fd0538cd1786d24f05421912555ac3/pkg/roachprod/roachprod.go#L1430)):*
<details><summary><i>Previously, herkolategan (Herko Lategan) wrote…</i></summary><blockquote>

It can and the idea did cross my mind. It's probably worth while.
</blockquote></details>

Nice! Now we can test how well `roachprod.Sync` handles concurrent invocations :)


<!-- Sent from Reviewable.io -->

herkolategan commented:
bors r=renatolabs,srosenberg
